### ISSUE TITLE
net/http: HTTP/1.1 performance degradation in 1.15beta1

### ISSUE THREAD

**Issue by proyb6** [golang/go#39543]
I am seeing 5% performance degradation when benchmark HTTP in plaintext on macOS Catalina. Ran test more than 3 times shown lesser 500 requests per second on 1.15beta1.

Simple hello world test on both 1 core:
go1.14.4 darwin/amd64
```
wrk -c100 -d5 http://localhost:8090/hello
Running 5s test @ http://localhost:8090/hello
  2 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.32ms    0.91ms  15.67ms   79.06%
    Req/Sec    15.16k     1.12k   18.43k    86.27%
  153782 requests in 5.10s, 19.07MB read
Requests/sec:  30156.31
Transfer/sec:      3.74MB
```

go1.15beta1 darwin/amd64
```
wrk -c100 -d5 http://localhost:8090/hello
Running 5s test @ http://localhost:8090/hello
  2 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.46ms    0.96ms  18.74ms   80.29%
    Req/Sec    14.55k     0.87k   17.22k    92.16%
  147583 requests in 5.10s, 18.30MB read
Requests/sec:  28941.41
Transfer/sec:      3.59MB
```

**Comment by ianlancetaylor** [comment]
Can you use the net/http/pprof package and collect CPU profiles from both versions?  Thanks.

CC @bradfitz

**Comment by proyb6** [comment]
[cpuprofile.zip](https://github.com/golang/go/files/4768870/cpuprofile.zip)

Attached on both versions.

**Comment by ianlancetaylor** [comment]
Thanks.  Unfortunately I don't see the problem.

I'll mark this as a release blocker for now.  I hope that somebody will be able to find what changed.

Can you share the code that you are using for benchmarking?  Thanks.

**Comment by proyb6** [comment]
Sure, the following example:

```
package main

import (
    "fmt"
    "net/http"
//  "runtime"
     _ "net/http/pprof"
)

func hello(w http.ResponseWriter, req *http.Request) {
        fmt.Fprintf(w, "hello world!\n")
}

func main() {
//      runtime.GOMAXPROCS(1)
        http.HandleFunc("/hello", hello)
        http.ListenAndServe(":8090", nil)
}
```

**Comment by hrsmntbygo** [comment]
# Go developers are harrassing me

- actively going against my Adsense sites
- actively attacking my github account
- actively deranking my page rank

You want war, you will get war, I promise you that

**Comment by andybons** [comment]
@neild

**Comment by neild** [comment]
I can confirm that I see a difference in 1.14 vs. 1.15beta1. Linux, amd64, GOMAXPROCS(1), the above test program and ten runs each of `wrk -c100 -d5`.

```
$ grep ^Request bench.go1.14
Requests/sec:  43433.19
Requests/sec:  43201.69
Requests/sec:  44116.82
Requests/sec:  43029.81
Requests/sec:  43827.57
Requests/sec:  42289.91
Requests/sec:  43815.39
Requests/sec:  44036.54
Requests/sec:  43955.58
Requests/sec:  43393.49

$ grep ^Request bench.go1.15
Requests/sec:  40063.74
Requests/sec:  40559.97
Requests/sec:  41783.38
Requests/sec:  41255.86
Requests/sec:  41534.14
Requests/sec:  40796.73
Requests/sec:  41519.93
Requests/sec:  39966.21
Requests/sec:  41283.03
Requests/sec:  41853.80
```

With default GOMAXPROCS, the difference was less pronounced. Nothing obvious to me in profiles. Will look further.

**Comment by dr2chase** [comment]
I took a stab at reproducing this just now, 1.14 vs 1.15tip, Linux, amd64, 12 processor box but GOMAXPROCS=1.  30 repetitions of 10 seconds of work:
```
( for i in {1..30} ; do wrk -c100 -d10 http://localhost:8114/hello; done; )  >& b114.log ; ( for i in {1..30} ; do wrk -c100 -d10 http://localhost:8115/hello; done; )  >& b115.log ;
```

Comparing averages, tip is 0.3% slower, comparing medians, 0.1% slower.

Got curious about what had changed since beta1, and noticed:
```
* 8b98498a58 net/http: make Transport.RoundTrip preserve Requests
* f84bbd52b0 cmd/compile: mark block control instruction unsafe in "all unsafe" functions
```

Benchmarked again at f84bbd52b0 (just before the change to net/http), and now it's 4% slower.

I will try the same experiment later today on a Mac laptop, once I am done using it for meetings.

**Comment by dr2chase** [comment]
I tried the comparison on a Mac laptop, Chrome and IDE shut down, but the numbers (30 repetitions of 10-second runs) were too noisy to be useful.  I need to try again, very carefully, but I it would be nice if the bug filer could try tip and see if things look better, since (once!) they seemed better for me on Linux.

**Comment by dmitshur** [comment]
Thanks for looking @dr2chase. I can try reproducing this on a Mac laptop too.

> it would be nice if the bug filer could try tip and see if things look better

@proyb6 Are you able to try again with the latest tip commit? (The [`gotip`](https://pkg.go.dev/golang.org/dl/gotip) command may be convenient for this.)

**Comment by dr2chase** [comment]
FYI, I ran a second experiment on Linux (same machine, but this time using "perflock" for the benchmarked server).
```
perflock ./114 >& 114-run.log &
for j in {1..5} ; do ( for i in {1..25} ; do wrk -c100 -d5 http://localhost:8114/hello; done; )  >& b114-${j}.log ; done;
kill %1

perflock ./115 >& 115-run.log &
for j in {1..5} ; do ( for i in {1..25} ; do wrk -c100 -d5 http://localhost:8115/hello; done; )  >& b115-${j}.log ; done;
kill %1

perflock ./115-tip >& tip-run.log &
for j in {1..5}
kill %1
```

First, it looks like the bug is indeed reproducible against f84bbd52b0.  Notice how there's not much difference among either group of 5, max/min is under 1%.

  | 114 run 1 | 114 run 2 | 114 run 3 | 114 run 4 | 114 run 5 | max/min |   | 115 run 1 | 115 run 2 | 115 run 3 | 115 run 4 | 115 run 5 | max/min
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  | 38399.48 | 38280.89 | 38339.71 | 38193.77 | 38316.30 | 1.005385715 |   | 36391.54 | 36522.23 | 36334.43 | 36273.72 | 36408.57 | 1.006850987
  | 38388.11 | 38315.11 | 38395.04 | 38400.81 | 38357.88 | 1.002236715 |   | 36563.58 | 36613.01 | 36353.43 | 36445.3 | 36562.06 | 1.007140454
  |   |   |   |   |   |   | mean(115)/mean(114) | 0.948 | 0.954 | 0.948 | 0.950 | 0.950 |  
  |   |   |   |   |   |   | median(115)/median(114) | 0.952 | 0.956 | 0.947 | 0.949 | 0.953 |  

It also looks like the bug is fixed at tip, and also that there's more noise among the runs at tip (same numbers for 114).  Note that I am taking maxs and mins of numbers that are already medians or averages, and I was using perflock, so this is a little interesting, maybe some service process woke up during the benchmarks.

  | 114 run 1 | 114 run 2 | 114 run 3 | 114 run 4 | 114 run 5 | max/min |   | tip run 1 | tip run 2 | tip run 3 | tip run 4 | tip run 5 | max/min
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
averages | 38399.48 | 38280.89 | 38339.71 | 38193.77 | 38316.30 | 1.005385715 |   | 38246.86 | 37958.87 | 38003.92 | 38281.55 | 38381.32 | 1.011129077
medians | 38388.11 | 38315.11 | 38395.04 | 38400.81 | 38357.88 | 1.002236715 |   | 38309.13 | 37897.57 | 38123.49 | 38318.6 | 38409.12 | 1.013498227
  |   |   |   |   |   |   | mean(tip)/mean(114) | 0.996 | 0.992 | 0.991 | 1.002 | 1.002 |  
  |   |   |   |   |   |   | median(tip)/median(114) | 0.998 | 0.989 | 0.993 | 0.998 | 1.001 |  

That is, I believe the bug is consistently reproducible on my pretty-quiet (office-quarantined) deskside box, and it has also been fixed, most likely by the first commit after f84bbd52b0 but perhaps by a later one.

**Comment by dmitshur** [comment]
I reproduced this with results similar to what's posted above on a macOS laptop, and I don't have much new information to add. The difference between 1.14.4 and tip is very small compared to the variance of the micro benchmark, and I wasn't able to narrow it down further.

It seems to correlate to the various changes to net/http between 1.14 and tip (https://github.com/golang/go/commits/master/src/net/http), most of which were minor strictness/correctness improvements, and none were optimizations.

Perhaps someone else can narrow it down further or more conclusively, but I did not find an actionable issue here.
